ic-utils/src/interfaces/http_request.rs

```diff
@@ -26,54 +28,239 @@ pub struct HttpRequestCanister;
 
 /// A key-value pair for a HTTP header.
 #[derive(Debug, CandidType, Clone, Deserialize)]
-pub struct HeaderField(pub String, pub String);
+pub struct HeaderField<'a>(pub Cow<'a, str>, pub Cow<'a, str>);
 
 /// The important components of an HTTP request.
-#[derive(Debug, Clone, CandidType, Deserialize)]
-pub struct HttpRequest<'body> {
+#[derive(Debug, Clone, CandidType)]
+struct HttpRequest<'a, H> {
     /// The HTTP method string.
-    pub method: String,
+    pub method: &'a str,
     /// The URL that was visited.
-    pub url: String,
+    pub url: &'a str,
     /// The request headers.
-    pub headers: Vec<HeaderField>,
+    pub headers: H,
     /// The request body.
-    #[serde(with = "serde_bytes")]
-    pub body: &'body [u8],
+    pub body: &'a [u8],
+}
+
+/// A wraper around an iterator of headers
+#[derive(Debug, Clone)]
+pub struct Headers<H>(H);
+
+impl<'a, H: Clone + ExactSizeIterator<Item = HeaderField<'a>>> From<H> for Headers<H> {
+    fn from(h: H) -> Self {
+        Headers(h)
+    }
+}
+
+impl<'a, H: Clone + ExactSizeIterator<Item = HeaderField<'a>>> CandidType for Headers<H> {
```

Comment:
Yes, first, we need `ExactSizeIterator` because candid prefixes the data with the size, meaning we need to call `len`. Second, `idl_serialize` takes a `&self`, meaning we need `Clone` to get the cheapest copy of the iterator.